src/indexer.c

```diff
@@ -154,52 +185,80 @@ RedisModuleDict * QueryIndexPredicate(RedisModuleCtx *ctx, QueryPredicate *predi
     RedisModuleString *index_key;
     size_t _s;
 
+    char *iter_ptr;
+    int nokey;
 
-    if (predicate->type == NCONTAINS || predicate->type == CONTAINS) {
-        index_key = RedisModule_CreateStringPrintf(ctx, K_PREFIX,
-                RedisModule_StringPtrLen(predicate->label.key, &_s));
+    const char *key = RedisModule_StringPtrLen(predicate->label.key, &_s);
+    const char *value;
 
-    } else {
-        const char *key = RedisModule_StringPtrLen(predicate->label.key, &_s);
-        const char *value = RedisModule_StringPtrLen(predicate->label.value, &_s);
-        index_key = RedisModule_CreateStringPrintf(ctx, KV_PREFIX, key, value);
-    }
-    int nokey;
-    currentLeaf = RedisModule_DictGet(labelsIndex, index_key, &nokey);
-    if (nokey) {
-        currentLeaf = NULL;
-    } else {
-        RedisModuleDictIter *iter = RedisModule_DictIteratorStartC(currentLeaf, "^", NULL, 0);
-        RedisModuleString *currentKey;
-        while((currentKey = RedisModule_DictNext(ctx, iter, NULL)) != NULL) {
-            RedisModule_DictSet(localResult, currentKey, (void *)1);
+    switch (predicate->type) {
+
+        case CONTAINS:
+        case NCONTAINS:
+            index_key = RedisModule_CreateStringPrintf(ctx, K_PREFIX,
+                                                       RedisModule_StringPtrLen(predicate->label.key, &_s));
+            break;
+        case NEQ:
+        case EQ:
+            value = RedisModule_StringPtrLen(predicate->label.value, &_s);
+            index_key = RedisModule_CreateStringPrintf(ctx, KV_PREFIX, key, value);
+            break;
+        default:
+            break;
+        }
+
+    if (predicate->type == FILTER_LIST) {
```

Comment:
@otnielvh the parsing should not be here, `QueryPredicate` should be changed to keep multi-value, and the parsing function for the predicates `parseLabelListFromArgs` (should be renamed and return/update a predicate)